### DISCUSSION TITLE
IfcGradientCurve modeling and implementation

### DISCUSSION THREAD

**Discussion by RickBrice** [IfcOpenShell/IfcOpenShell#5587] [General]
I would like to have a discussion about modeling and implementation of IfcGradientCurve. This discussion stems from the models for Issue #5502.

For the attached models, the vertical profile is as shown below. The horizontal composite curve is a single line segment from (0,0) to (120,0).

![image](https://github.com/user-attachments/assets/6b5118d9-183d-466f-a7f0-d57f3587a54f)

[github_issue_5502_example_4.ifc.txt](https://github.com/user-attachments/files/17400686/github_issue_5502_example_4.ifc.txt)
[github_issue_5502_example_5.ifc.txt](https://github.com/user-attachments/files/17400685/github_issue_5502_example_5.ifc.txt)

The difference between the "example_4" and "example_5" models are the IfcLine parent curves. For example_4, the line is horizontal. For example_5, the line is sloped. All IfcCurveSegment in the IfcGradientCurve use the same IfcLine parent curve.

example_4
~~~
#59=IFCCARTESIANPOINT((0.,0.));
#60=IFCDIRECTION((1.,0.));
#61=IFCVECTOR(#60,1.);
#62=IFCLINE(#59,#61);
~~~

example_5
~~~
#59=IFCCARTESIANPOINT((0.,0.));
#60=IFCDIRECTION((0.995037190209989,-0.0995037190209989));
#61=IFCVECTOR(#60,1.);
#62=IFCLINE(#59,#61);
~~~

It is (or was) my understanding that the location and rotation of the parent curve is irrelevant. The IfcCurveSegment trims a portion of the parent curve. Then the trimmed curve is placed such that its start point and tangent are aligned with the IfcCurveSegment.Placement location and tangent.

To evaluate any point on a segment of the gradient curve, the point is computed on the parent curve, and then translated and rotated using the curve segment placement.

Both models are evaluated at "u = 10.0" along the horizontal. The result is the same in both cases.
[[ 1.  0.  0. 10.]
 [ 0.  1.  0.  0.]
 [ 0.  0.  1. -2.]
 [ 0.  0.  0.  1.]]

This implies the parent curve rotation doesn't matter.

Evaluating example_4 at "u = 30.0", the result is
[[ 0.99503719  0.          0.09950372 30.        ]
 [ 0.          1.          0.          0.        ]
 [-0.09950372  0.          0.99503719 -2.99503719]
 [ 0.          0.          0.          1.        ]]

However, example_5 at "u = 30.0", the result is:
[[ 0.99503719  0.          0.09950372 30.        ]
 [ 0.          1.          0.          0.        ]
 [-0.09950372  0.          0.99503719 -3.        ]
 [ 0.          0.          0.          1.        ]]

example_5 gives the correct result. This implies the rotation of the parent curve is important.

So, my questions are:
1. Is the parent curve rotation important?
2. Is the example_4 model valid?
3. In the example_5 model, should the trim length for IfcCurveSegment 66, 82, 98 be 20.0 (distance along horizontal) or 20.0/0.995037=20.099755 (distance along the sloped curve segment)?

I have managed to get myself very confused. Discussion is appreciated.
@aothms @civilx64 @peterrdf @SergejMuhic @jmirtsch @JanErikHoel

**Comment by jmirtsch** [comment]
I took a look at example 5 model.  I don't believe it's right.  I believe as you suggest, the gradient segment length should be the "arc' length along the segment, not in the case of a gradient curve, the "horizontal" length.

As I understand it, the intention is for the extent of the gradient curve to run the entire extent of the horizontal curve (not sure why the last gradient segment isn't explicitly 0 long, but the placement is at 120m along the horizontal curve which is the entire length).

I recreated the example curve from #5502 , you can find this here.  
[241017 alignment.ifc.txt](https://github.com/user-attachments/files/17406065/241017.alignment.ifc.txt)
So my understanding to respond to your questions.
1.  The location and rotation of the segment parent curve is not important.  This does enable efficiency as the same curve can be reused for multiple segments.
2.  I don't believe example 4 is valid, the sloping segments should have a length greater than 20.  You can compare with my understanding in the example above.
3. I suggest that the trim length is greater than 20 for the sloping segments.

I welcome any feedback if my understanding isn't correct.

**Comment by RickBrice** [comment]
Thanks @jmirtsch. I agree that the example 4 and 5 models are incorrect and the gradient segment length trimmed from the parent curve is the 'arc' length which would be a bit more than 20 (about 20/0.955 = 20.942).

Since the location and rotation of the segment parent curve is not important, then I have a misunderstanding about how to do the math.

For simplicity, I took your example and changed the alignment to a straight line from (0,0) to (120,0). This will simplify the math.
[241017.alignment.revised.ifc.txt](https://github.com/user-attachments/files/17414159/241017.alignment.revised.ifc.txt)

I want to compute (X,Y,Z) at a distance along the horizontal alignment of 30.0. The result should be (30,0,-3).

Here is how I do the calculations.
1) Evaluate parent curve at (30-20) = 10 from the start of the second gradient curve segment
The parent curve is a horizontal line, (x.y) = (0,0), (dx,dy) = (1,0). The resulting matrix is:
~~~
X = x + (dx/dx)*u = 0 + (1/1)*10 = 10
Y = y + (dy/dx)*u = 0 + (0/1)*10 = 0

[ dx,dy,0,X]   [1,0,0,10]
[-dy,dx,0,Y] = [0,1,0, 0]
[  0, 0,1,0]   [0,0,1, 0]
[  0, 0,0,1]   [0,0,0, 1]
~~~

2) Remove parent curve location and rotation so the computed point is relative to (0,0) with tangent direction (1,0). Since this is already the case, no calculation needed.

3) Apply the curve segment placement by matrix multiplication (this is for the second gradient curve segment)
~~~
[0.955,  0.0955, 0, 20]     [1,0,0,10]     [0.955,   0.0955, 0,  29.95]
[-0.0955, 0.955, 0, -2]  *  [0,1,0, 0]  =  [-0.0955, 0.955,  0, -2.995]
[0,       0,     1,  0]     [0,0,1, 0]     [0,       0,      1,  0    ]
[0,       0,     0,  1]     [0,0,0, 1]     [0,       0,      0,  1    ]
~~~

4) Compute coordinate for horizontal
~~~
[1,0,0,30]
[0,1,0,0]
[0,0,1,0]
[0,0,0,1]
~~~

5) Combine horizontal and vertical
Remove "X" from vertical result so it doesn't add to "X" from horizontal (i.e., set 29.95 from step 3 to 0.0)
Swap row 2 and 3 and column 2 and 3 in vertical result, because in the gradient curve Y is elevation and in 3D, Z is elevation.
Revised vertical result matrix is:
~~~
[0.955,   0, 0.0955, 0    ]
[0,       1,      0, 0    ]
[-0.0955, 0,  0.955, -2.995]
[0,       0,      0, 1    ]
~~~
Multiply horizontal and vertical to get final result
~~~
[1,0,0,30]    [0.955,   0, 0.0955,     0     ]     [0.955,   0, 0.0955,     30.0  ]
[0,1,0, 0]  * [0,       1,      0,     0     ]  =  [0,       1,      0,     0     ]
[0,0,1, 0]    [-0.0955, 0.,     0.955, -2.995]     [-0.0955, 0.,     0.955, -2.995]
[0,0,0, 1]    [0,       0,      0,      1    ]     [0,       0,      0,     1     ]
~~~
The resulting point is (30,0,-2.995). **This is definitely the wrong result**.

However, if the parent curve is sloped (dx,dy) = (0.955,-0.0955) the result of step 1 is
~~~
X = x + (dx/dx)*u = 0 + (0.955/0.955)*10 = 10
Y = y + (dy/dx)*u = 0 + (-0.0955/0.955)*10 = -1

[0.955,  0.0955,0,10]
[-0.0955,0.955, 0,-1]
[0,      0,     1, 0]
[0,      0,     0, 1]
~~~

In step 2, the parent curve rotation is removed
~~~
[0.955,  -0.0955, 0, 0]   [1,0,0,10]    [1,0,0,10.0495]
[0.0955,  0.955,  0, 0] * [0,1,0,-1] =  [0,1,0,-1]
[0,       0,      1, 0]   [0,0,1, 0]    [0,0,1, 0]
[0,       0,      0, 1]   [0,0,0, 1]    [0,0,0, 1]
~~~

Then step 3, apply the curve segment placement
~~~
[0.955,  0.0955, 0, 20]     [1,0,0,10.0495]      [0.955,   0.0955, 0, 30]
[-0.0955, 0.955, 0, -2]  *  [0,1,0,0]         =  [-0.0955, 0.955,  0, -3]
[0,       0,     1,  0]     [0,0,1,0]            [0,       0,      1,  0]
[0,       0,     0,  1]     [0,0,0,1]            [0,       0,      0,  1]
~~~

Swap rows and columns of the vertical matrix and combine with horizontal.
The final result is
~~~
[0.955,   0, 0.0955, 30]
[0,       1,      0, 0 ]
[-0.0955, 0,  0.955, -3]
[0,       0,      0, 1 ]
~~~
This is the correct result.

For my implementation, the rotation of the parent curve matters - this tells me I've got something fundamentally wrong.

Any ideas where my mistake is?

**Comment by civilx64** [comment]
I don't have much to contribute yet but wanted to announce that I'm lurking and investigating the implementation in IfcOpenShell for an upcoming best practice rule ALS017 for the Validation Service regarding geometric continuity between IfcGradientCurve segments.

**Reply by RickBrice** [reply]
Have you tried the example 4 and 5 files against the ALS017 rule? I would expect errors from both since the length of the vertical sloped segments are too short.

**Reply by civilx64** [reply]
I have not tried examples 4 and 5 yet but the model from @jmirtsch is passing ALS017:

![image](https://github.com/user-attachments/assets/1107c70c-9e00-415c-a63b-de6322ef6395)
![image](https://github.com/user-attachments/assets/b9a159d9-3704-471d-a015-db6b197e2321)

**Reply by civilx64** [reply]
examples 4 and 5 do not contain any alignments, so they are not activated by ALS017 ( do not meet the selection criteria of the second step).

**Comment by civilx64** [comment]
> I suggest that the trim length is greater than 20 for the sloping segments.

I agree.

Doing some investigation of the FHWA Bridge Geometry Example:

Initially I was seeing some differences at the end of the first segment (CONSTANTGRADIENT) / beginning of the 2nd segment (PARABOLICARC).

 _Note: ifcopenshell.version_core = '0.7.11-d51fa2c5f'_

```
Calculated end of IfcCurveSegment(120) = (1199.8162921943099, 120.99678511340042)
Observed start point of IfcCurveSegment(129) = (1200.0, 121.0)
```

This was with SegmentLength = 1200. as the distance from Station 100+00 to 112+00.  After adjusting SegmentLength to 1200.1873593 things look better:

```
Calculated end of IfcCurveSegment(120) = (1199.9999999962063, 120.9999999999336)
Observed start point of IfcCurveSegment(129) = (1200.0, 121.0)
```

Granted this is all still in the domain of (dist_along, height) of the vertical alignment and doesn't address your question above regarding full XYZ coordinates of the alignment at the segment interface points.  That will need to be another day, at least for me...

**Reply by RickBrice** [reply]
> This was with SegmentLength = 1200. as the distance from Station 100+00 to 112+00. After adjusting SegmentLength to 1200.1873593 things look better:
> doesn't address your question above

I think this is very related to my question. I can see changing the IfcCurveSegment.SegmentLength to 1200.1873593 because this is the distance along the line that needs to be trimmed. Now carry this same logic to the next curve segment which is a parabola. IfcCurveSegment.SegmentLength = 1600.0. This is the horizontal distance, not the length to be trimmed from the parabola.

Is this the same issue @jmirtsch brought up in recent emails?

**Comment by RickBrice** [comment]
I think I understand the error in my thinking. IfcCircle is defined by a placement and a radius. IfcLine is defined by a point and direction. The axes of the coordinate system defining a circle can be rotated. The axes for a line are fixed horizontal and vertical. Because of this, the Direction of the parent curve IfcLine is important. For vertical, the direction of the line defines Y given a horizontal distance X.

The short answer is, the IfcGradientCurve in some of my files are wrong because (1) the direction of IfcLine must represent the vertical slope, and (2) the trim length of the segment should be length along the line, not the horizontal length.

So, for the example provided by @jmirtsch - There seems to be an error that the validation service isn't detecting. There should be two different IfcLine parent curves for the segments of the gradient curve. One IfcLine.Direction = (1.,0.) for the horizontal segments and IfcLine.Direction = (0.955,-0.0955) for the sloped segments.

This is the model Jon provided, but modified in the following way. The alignment is horizontal and 1200.0 long. All gradient curve points have been scaled by 10 in x and y. When evaluating the gradient curve at 300.0 from the start (instead of 30), the result is Y = 29.95 (instead of 2.995). The correct value is 30.0 (instead of 3.0) so the difference should now be big enough for the VS rule to pick it up. 

@civilx64 can you run the attached file?

[241017.alignment.revised.ifc.txt](https://github.com/user-attachments/files/17437237/241017.alignment.revised.ifc.txt)

**Reply by jmirtsch** [reply]
I don't understand the suggestion (1) that the direction of IfcLine must represent the vertical slope.  This is counter to all discussions I recall from when the standard was under going testing.
The resultant trimmed segment is placed in the context of the gradient base curve by the placement.  Here I highlight where the grade of the sloping segments is nominated in my example.  The documentation might need improvement, so this is clearer.
It is legitimate that the parent curve does have the correct slope though, so both can work.

![2024-10-19_06-51-44](https://github.com/user-attachments/assets/2a1b7433-730c-496c-b97b-d415e2774a72)


The start tangent of the trimmed segment needs to be determined, and then aligned with the segment placement.
I only count one occurrence of "IfcLine" if your modified file.  So it doesn't seem consistent with your suggestion.
I'd like clarity on this if you do believe my file is not valid.

Here is an updated file with a single straight segment in the horizontal, with length 1200.

[241019 alignment.ifc.txt](https://github.com/user-attachments/files/17440365/241019.alignment.ifc.txt)

This doesn't pass validation only for the lack of at least one IfcReferent (which seems to be a new test).  I'm not sure why an alignment needs at least one referent if no stationing is nominated etc.

**Reply by civilx64** [reply]
> @civilx64 can you run the attached file?

It still passes, which I believe is correct.

> There should be two different IfcLine parent curves for the segments of the gradient curve. One IfcLine.Direction = (1.,0.) for the horizontal segments and IfcLine.Direction = (0.955,-0.0955) for the sloped segments.

Here is where I think you're getting turned around.  Direction of (0.995, -0.0995) is for the RefDirection of the Placement for vertically sloping segments as noted by @jmirtsch above.  Direction of (1., 0.) is for the ref direction of the Placement for flat segments.  This is not the direction of the IfcLine ParentCurve that provides the "whole" that gets trimmed based on length before inserting into the IfcGradientCurve.

**Reply by civilx64** [reply]
@jmirtsch 


![image](https://github.com/user-attachments/assets/1a745b02-105b-4bcd-8044-eb3f8c67dd55)


This result should be understood as "passing".  [ALB010 is a new addition](https://forums.buildingsmart.org/t/validation-service-updates/4992/12) indicating that associating at least one IfcReferent to an alignment is a best practice, but not a requirement.

Perhaps the wording could be improved in 
![image](https://github.com/user-attachments/assets/fd3f3606-89ea-490e-b67f-15a47bdbd8dd)

by revising "must exist" to "should exist". Thoughts?

**Reply by civilx64** [reply]
@RickBrice 

> Is this the same issue @jmirtsch brought up in recent emails?

Not exactly; that discussion references the fact that IfcCurveSegment.SegmentStart and IfcCurveSegment.SegmentLength are of type IfcCurveMeasureSelect which includes types IfcLengthMeasure and IfcParameterValue.  IfcCurveSegment has a documented Informal Proposition that these attributes shall be of type IfcLengthMeasure, which is enforced by the Valdiation Service via rule ALS012.

**Reply by civilx64** [reply]
@jmirtsch Sorry, one follow-up clarification:

> It is legitimate that the parent curve does have the correct slope though, so both can work.

> The start tangent of the trimmed segment needs to be determined, and then aligned with the segment placement.

If I understand this correctly, an alternative approach for this model would be to have two IfcLine entities available for use as ParentCurves:
* one with direction (1., 0.) for flat vertical segments
* one with direction (0.955, -0.0955) for the vertically sloping segments

And then for each IfcCurveSegment, reference the appropriate IfcLine and use (0., 0.) as RefDirection for the Placement.

Do I have that correct?

**Reply by aothms** [reply]
Just tangentially following along, but:

>  It is legitimate that the parent curve does have the correct slope though, so both can work.

Since we're still 'early' in standardization, and can nudge people towards consistent behaviour with the VS, how about we make a choice here? That we require that line direction always needs to be (1, 0), position always (0, 0) and the placement is used for positioning (or the other way around if that's more flexible / semantically fit).

**Comment by SergejMuhic** [comment]
Some background on `IfcCurveSegment`. It was introduced because alignment is typically a lot of segments and `IfcTrimmedCurve` that was available is somewhat verbose. This is why `IfcCurveSegment` has some inherent propositions to make it easier, faster and more space efficient to implement.

As @jmirtsch explained, the rationale was simplicity and since the inherent traits of `IfcCurveSegment` deal with both orientation and position, it was agreed that related attributes on `IfcCurveSegment.ParentCurve` shall be ignored. We even looked at making something like `IfcConic.Position` optional. Due to backwards compatibility, this was reverted eventually.

**Comment by RickBrice** [comment]
Everyone, thank you for the discussion. I got myself turned around with parent curve definitions and the curve segment. This has lead me to do the math incorrectly for vertical profiles. I think I can get it straightened out. @jmirtsch i think your file is correct. It’s my misunderstanding of the parent curve. 

I’ll try to get IfcOpenShell fixed up quickly. In the interim, Scott vertical gradients may not be evaluated correctly in all cases. The error magnifies with length of the gradient and the steepness of the slope. This will likely affect the VS rule you are working on.

**Comment by RickBrice** [comment]
@civilx64 I got this fixed so vertical gradient segments should evaluate correctly regardless of the slope of the parent curve line.
